### PR TITLE
GODRIVER-1450 : use correct writeModel in BulkWriteError

### PR DESCRIPTION
**Pull request by emilien-puget**

### CODE REVIEW

File: mongo/bulk_write.go [hunk 1]
```diff
@@ -151,9 +151,14 @@ func (bw *bulkWrite) runBatch(ctx context.Context, batch bulkWriteBatch) (BulkWr
 	batchErr.WriteErrors = make([]BulkWriteError, 0, len(writeErrors))
 	convWriteErrors := writeErrorsFromDriverWriteErrors(writeErrors)
 	for _, we := range convWriteErrors {
+		model := batch.models[0]
+		if len(batch.models) > we.Index {
```
Comment (divjotarora) [inline mongo/bulk_write.go]: Is this check necessary? Is it possible to have a case where the index is nit valid for `batch.models`?

Comment (emilien-puget) [reply mongo/bulk_write.go]: i don't think so, but as I could not find anything about that on the documentation, i prefered to be safer

Comment (divjotarora) [reply mongo/bulk_write.go]: The index should always be the index in the current batch, so I think it's fine to have `batch.models[we.Index]` unconditionally.

**Comment by emilien-puget** [comment]
i could not find where the integration tests are, if you can point it to me i can add one
my local test consist if doing a bulk write like this

```
[]mongo.WriteModel{
		mongo.NewUpdateOneModel().SetFilter(bson.M{}).SetUpdate(bson.M{}),
		mongo.NewInsertOneModel().SetDocument(bson.M{
			"email": "notduplicate@test.com",
		}),
		mongo.NewInsertOneModel().SetDocument(bson.M{
			"email": "duplicate1@test.com",
		}),
		mongo.NewInsertOneModel().SetDocument(bson.M{
			"email": "duplicate1@test.com",
		}),
		mongo.NewInsertOneModel().SetDocument(bson.M{
			"email": "duplicate2@test.com",
		}),
		mongo.NewInsertOneModel().SetDocument(bson.M{
			"email": "duplicate2@test.com",
		}),
	}
```

on a collection with an index like this
```
mongo.IndexModel{
		Keys: bson.D{
			bson.E{Key: "email", Value: 1},
		},
		Options: options.Index().
			SetUnique(true),
	})
```

the error should be a `BulkWriteException`

if the bulkwrite was set to ordered at false you should have one element in the BulkWriteException, otherwise 2

but the missing thing is that on those elements, the Request property should be, in the case of ordered false, 
```
bson.M{
			"email": "duplicate1@test.com",
		}
```

it is instead
```
bson.M{
			"email": "notduplicate@test.com",
		}
```

File: mongo/bulk_write.go (hunk 1 above)
Comment (emilien-puget) [reply mongo/bulk_write.go]: all right, i made another comment describing my steps to test that

**Comment by divjotarora** [comment]
You can write a test in `mongo/integration/collection_test.go` starting right after https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/collection_test.go#L966. We use a custom integration testing framework that sets up the proper client/collection objects, so the test will be something like:

```
mt.Run("correct model in errors", func(mt *mtest.T) {
			models := []mongo.WriteModel{
				mongo.NewUpdateOneModel().SetFilter(bson.M{}).SetUpdate(bson.M{}),
				mongo.NewInsertOneModel().SetDocument(bson.M{
					"_id": "notduplicate",
				}),
				mongo.NewInsertOneModel().SetDocument(bson.M{
					"_id": "duplicate1",
				}),
				mongo.NewInsertOneModel().SetDocument(bson.M{
					"_id": "duplicate1",
				}),
				mongo.NewInsertOneModel().SetDocument(bson.M{
					"_id": "duplicate2",
				}),
				mongo.NewInsertOneModel().SetDocument(bson.M{
					"_id": "duplicate2",
				}),
			}

			_, err := mt.Coll.BulkWrite(mtest.Background, models)
			bwException, ok := err.(mongo.BulkWriteException)
			assert.True(mt, ok, "expected error of type %T, got %T", mongo.BulkWriteException{}, err)

			expectedModel := models[3]
			actualModel := bwException.WriteErrors[0].Request
			assert.Equal(mt, expectedModel, actualModel, "expected model %v in BulkWriteException, got %v",
				expectedModel, actualModel)
		})
```

Apologies for the poor spacing, but that should do it. The only difference between this and your test is that I used the `_id` field rather than `email` to avoid creating a new index in the test.

**Comment by emilien-puget** [comment]
> You can write a test in `mongo/integration/collection_test.go` starting right after https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/collection_test.go#L966. We use a custom integration testing framework that sets up the proper client/collection objects, so the test will be something like:
> 
> ```
> mt.Run("correct model in errors", func(mt *mtest.T) {
> 			models := []mongo.WriteModel{
> 				mongo.NewUpdateOneModel().SetFilter(bson.M{}).SetUpdate(bson.M{}),
> 				mongo.NewInsertOneModel().SetDocument(bson.M{
> 					"_id": "notduplicate",
> 				}),
> 				mongo.NewInsertOneModel().SetDocument(bson.M{
> 					"_id": "duplicate1",
> 				}),
> 				mongo.NewInsertOneModel().SetDocument(bson.M{
> 					"_id": "duplicate1",
> 				}),
> 				mongo.NewInsertOneModel().SetDocument(bson.M{
> 					"_id": "duplicate2",
> 				}),
> 				mongo.NewInsertOneModel().SetDocument(bson.M{
> 					"_id": "duplicate2",
> 				}),
> 			}
> 
> 			_, err := mt.Coll.BulkWrite(mtest.Background, models)
> 			bwException, ok := err.(mongo.BulkWriteException)
> 			assert.True(mt, ok, "expected error of type %T, got %T", mongo.BulkWriteException{}, err)
> 
> 			expectedModel := models[3]
> 			actualModel := bwException.WriteErrors[0].Request
> 			assert.Equal(mt, expectedModel, actualModel, "expected model %v in BulkWriteException, got %v",
> 				expectedModel, actualModel)
> 		})
> ```
> 
> Apologies for the poor spacing, but that should do it. The only difference between this and your test is that I used the `_id` field rather than `email` to avoid creating a new index in the test.

all right, its is better like that

**Comment by divjotarora** [comment]
Thanks so much for the contribution @emilien-puget! I've merged and backported this fix and it will be released in driver version 1.2.1.

-- Divjot

**Comment by emilien-puget** [comment]
Hello, is it also possible to put this fix in the 1.1.X ?

**Comment by divjotarora** [comment]
Hi @emilien-puget,

We only backport to the latest minor version, which is currently the 1.2.x branch, unless there are special circumstances (i.e. we think the issue is severe and might affect multiple customers in production).

**Comment by emilien-puget** [comment]
alright, thank you. Do you have an estimate to when this new version will be released ?

**Comment by divjotarora** [comment]
@emilien-puget The plan is to release 1.2.1 early this week.

**Comment by divjotarora** [comment]
@emilien-puget Just wanted to let you know that the new version has been released and is available under the v1.2.1 tag.

**Comment by emilien-puget** [comment]
Thank you very much for letting me know

**Pull request merged**